src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_evp_pkey_rsa.c

```diff
@@ -32,6 +35,81 @@ EVP_PKEY* CryptoNative_RsaGenerateKey(int keySize)
     return ret;
 }
 
+int32_t CryptoNative_RsaDecrypt(EVP_PKEY* pkey,
+                                const uint8_t* source,
+                                int32_t sourceLen,
+                                RsaPaddingMode padding,
+                                const EVP_MD* digest,
+                                uint8_t* destination)
+{
+    assert(pkey != NULL);
+    assert(source != NULL);
+    assert(destination != NULL);
+    assert(padding >= RsaPaddingPkcs1 && padding <= RsaPaddingOaepOrPss);
+    assert(digest != NULL || padding == RsaPaddingPkcs1);
+
+    EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new(pkey, NULL);
+
+    int ret = -1;
+
+    if (ctx == NULL || EVP_PKEY_decrypt_init(ctx) <= 0)
+    {
+        goto done;
+    }
+
+    if (padding == RsaPaddingPkcs1)
+    {
+        if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING) <= 0)
+        {
+            goto done;
+        }
+    }
+    else
+    {
+        assert(padding == RsaPaddingOaepOrPss);
+
+        if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_OAEP_PADDING) <= 0)
+        {
+            goto done;
+        }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wcast-qual"
+        if (EVP_PKEY_CTX_set_rsa_oaep_md(ctx, digest) <= 0)
+#pragma clang diagnostic pop
+        {
+            goto done;
+        }
+    }
+
+    // This check may no longer be needed on OpenSSL 3.0
+    {
+        RSA* rsa = EVP_PKEY_get0_RSA(pkey);
+
+        if (rsa == NULL || HasNoPrivateKey(rsa))
+        {
+            ERR_PUT_error(ERR_LIB_RSA, RSA_F_RSA_NULL_PRIVATE_DECRYPT, RSA_R_VALUE_MISSING, __FILE__, __LINE__);
+            ret = -1;
+            goto done;
+        }
+    }
+
+    size_t written;
+
+    if (EVP_PKEY_decrypt(ctx, destination, &written, source, Int32ToSizeT(sourceLen)) > 0)
```

Comment:
I think `written` needs to be initialized to the length of `destination` here. From OpenSSL docs:

> If out is not NULL then before the call the outlen parameter should contain the length of the out buffer, if the call is successful the decrypted data is written to out and the amount of data written to outlen.

So `written` is an in/out, not just an out and might be garbage. OpenSSL then checks the that the `outlen` input is long enough to write to.

The validation call is here:

https://github.com/openssl/openssl/blob/52c587d60be67c337364b830dd3fdc15404a2f04/crypto/evp/pmeth_fn.c#L200

And if `arglen` (written in your case) is too small after checking it against the key size, it throws.

https://github.com/openssl/openssl/blob/52c587d60be67c337364b830dd3fdc15404a2f04/crypto/evp/pmeth_fn.c#L29-L30

If you initialize `written` to zero first, the error will happen consistently, just to demonstrate that the value prior to calling `EVP_PKEY_decrypt` matters.

Reply:
I opened a PR into your PR for this here. https://github.com/bartonjs/runtime/pull/1

If the PR into a PR is too weird or doesn't work you can grab a patch here and apply it directly to your branch https://github.com/bartonjs/runtime/pull/1.patch (assuming the changes are useful :smile: )

Reply:
Oy, what an embarrassing RTFM moment.  Thanks for the detective work.

Reply:
> I opened a PR into your PR for this here

Huh, the tab had gotten your first message, but not the second.  I didn't ignore your PRPR, just happened to have made a pretty-darn-close-to-identical version before seeing that it existed (when catching up on email notifications).

Reply:
No worries. I'm glad we pretty much came to the same change set. 😀

Reply:
Yay, that seemed to do it.